api/src/oscc.c

```diff
@@ -492,110 +603,372 @@ oscc_result_t oscc_async_enable( int socket )
     return result;
 }
 
-oscc_result_t oscc_init_can( const char *can_channel )
+
+oscc_result_t oscc_search_can( struct can_contains(*search_callback)( const char * ),
+                               bool search_oscc )
 {
-    int result = OSCC_ERROR;
-    int ret = -1;
+    oscc_result_t result = OSCC_ERROR;
 
+    static struct device_name *dev_list, *temp_ptr;
+    dev_list = malloc( sizeof( struct device_name ) );
 
-    struct sigaction sig;
+    result = construct_interfaces_list( dev_list );
 
-    memset( &sig, 0, sizeof(sig) );
-    sigemptyset( &sig.sa_mask );
-    sig.sa_flags = SA_RESTART;
-    sig.sa_handler = oscc_update_status;
-    sigaction( SIGIO, &sig, NULL );
+    //temp_contents is the temporary storage of the current CAN channel
+    //all_contents is the sum of all channels searched
+    struct can_contains temp_contents, all_contents;
 
-    int sock = socket( PF_CAN, SOCK_RAW, CAN_RAW );
+    //Set the all_contents based search_oscc boolean in function call
+    all_contents.is_oscc = !search_oscc;
+    all_contents.has_vehicle = false;
 
-    if ( sock < 0 )
+    temp_ptr = dev_list;
+
+    do
     {
-        printf( "Opening CAN socket failed: %s\n", strerror(errno) );
+        if ( strstr( temp_ptr->name, "can" ) != NULL )
+        {
+            temp_contents = search_callback( temp_ptr->name );
+
+            all_contents.is_oscc |= temp_contents.is_oscc;
+
+            all_contents.has_vehicle |= temp_contents.has_vehicle;
+
+            //Leave the while loop if both requirements are met
+            if( all_contents.is_oscc && all_contents.has_vehicle )
+            {
+                break;
+            }
+        }
+
+        temp_ptr = temp_ptr->next;
```

Comment:
This appears to be the only place where we traverse the `device_name` linked list, and we only follow the `next` pointer. Given that we don't use the `prev` pointer can we omit that pointer to simplify things?